### DISCUSSION TITLE
ThreadDataloader GIL

### DISCUSSION THREAD

**Discussion by hw-ju** [Project-MONAI/MONAI#4158] [Q&A] [answered]
Hi! In the situation where we cache all deterministic transforms in GPU memory and then use ThreadDataloader to perform lightweight non-deterministic GPU transforms, does the separate thread used by ThreadDataloader release GIL so that it doesn't blocker other threads which are running training? Thanks!

**Comment by Nic-Ma** [comment]
Hi @ericspod ,

Could you please help share some comments here?

Thanks in advance.

**Comment by ericspod** [comment, marked as answer]
The `ThreadDataLoader` class is using Python threads which still need to have the GIL locked to execute. If all your transforms are pure Python then this will interfere with the main (training) thread, however typically transforms are using Numpy, Pytorch, or some other library using compiled code which releases the GIL. The purpose of this class is to take advantage of this and of the fact that training is mostly done in compiled code, despite the GIL it will in certain situations allow overlapping of training and batch queuing. If either training time or batch queue time a much longer than the other there will be less benefit however.

**Reply by hw-ju** [reply]
Thanks a lot @ericspod @Nic-Ma! Crystal clear explanation! :D
